### PR TITLE
Min max preferred amount

### LINKED ISSUE #18: Updating the SolidusAftepay::PaymentMethod sets the minimum_amount and maximum_amount preferences to 0.0
Updating the `SolidusAftepay::PaymentMethod` sets the `minimum_amount` and `maximum_amount` preferences to `0.0` instead of leaving them `nil`. This is causing an issue [here](https://github.com/nebulab/solidus_afterpay/blob/main/app/models/solidus_afterpay/payment_method.rb#L42-L43) because instead of retrieving the configuration from Afterpay, the method always returns  a `0.0..0.0` range.

### PR DESCRIPTION
**Pull request by garciajordy**
# Min/Max Amount Bug Fix

Add cache methods for currency, minimum and maximum amount values when fetching this data from Afterpay with an expiry value of 1 day as this will slow down the performance of every user that tries to checkout have to make an API call.

fix #18

### CODE REVIEW

File: app/models/solidus_afterpay/payment_method.rb [hunk 1]
```diff
@@ -33,24 +31,39 @@ def try_void(payment)
...
 
-    def configuration
-      @configuration ||= gateway.retrieve_configuration
+    def maximum_amount
+      Rails.cache.fetch("solidus_afterpay_max_amount", expires_in: expires_in_seconds) do
+        configuration&.maximumAmount&.amount.to_f
+      end
+    end
+
+    def currency
```
Comment (ChristianRimondi) [inline app/models/solidus_afterpay/payment_method.rb:59]: why did you rename this method from `available_payment_currency` to `currency`?

File: app/models/solidus_afterpay/payment_method.rb [hunk 2]
```diff
@@ -33,24 +31,39 @@ def try_void(payment)
...
+      Rails.cache.fetch("solidus_afterpay_currency", expires_in: expires_in_seconds) do
```
Comment (ChristianRimondi) [inline app/models/solidus_afterpay/payment_method.rb:60]: You can just use `SolidusAfterpay.config.cache_expiry` here

File: spec/models/solidus_afterpay/payment_method_spec.rb [hunk 3]
```diff
@@ -140,4 +140,29 @@
...
+    end
+
+    it "retrieves the minimum amount from cache" do
+      cache.write("solidus_afterpay_min_amount", 1.0)
+      expect(described_class.new.send(:minimum_amount)).to eq(1.0)
+    end
+
+    it "retrieves the maximum amount from cache" do
+      cache.write("solidus_afterpay_max_amount", 1000.0)
+      expect(described_class.new.send(:maximum_amount)).to eq(1000.0)
```
Comment (ChristianRimondi) [inline spec/models/solidus_afterpay/payment_method_spec.rb:160]: it's now good practice to test `private` methods. You should test the `available_for_order?` method with different caching scenarios instead (with and without cache hit).

File: app/models/solidus_afterpay/payment_method.rb [hunk 4]
```diff
@@ -6,9 +6,7 @@ class PaymentMethod < SolidusSupport.payment_method_parent_class
     preference :secret_key, :string
     preference :deferred, :boolean
     preference :popup_window, :boolean
-    preference :minimum_amount, :decimal
```
Comment (ChristianRimondi) [inline app/models/solidus_afterpay/payment_method.rb:9]: here I think it's better to remove these preferences in a different commit. And remember to remove them also in the spec 😄

File: app/models/solidus_afterpay/payment_method.rb [hunk 5]
```diff
@@ -33,24 +31,39 @@ def try_void(payment)
...
-      min..max
+    def configuration
+      @configuration ||= gateway.retrieve_configuration
     end
 
-    def available_payment_currency
-      preferred_currency || configuration&.maximumAmount&.currency
+    def minimum_amount
+      Rails.cache.fetch("solidus_afterpay_min_amount", expires_in: expires_in_seconds) do
+        configuration&.minimumAmount&.amount.to_f
```
Comment (ChristianRimondi) [inline app/models/solidus_afterpay/payment_method.rb:49]: We should find a way to handle `configuration` errors. The `retrieve_configuration` method returns `nil` in case of Afterpay errors (e.g. request timeout), and invalid values will be cached for one day. Maybe we can use something like this (untested):
```ruby
Rails.cache.fetch("solidus_afterpay_min_amount", expires_in: expires_in_seconds) do
  configuration&.minimumAmount&.amount.to_f
end.tap do |_|
  Rails.cache.delete('solidus_afterpay_min_amount') if configuration.nil?
end
```

@vassalloandrea WDYT?

**Review by ChristianRimondi — CHANGES_REQUESTED** [review]

File: app/models/solidus_afterpay/payment_method.rb (hunk 1 above)
Comment (garciajordy) [reply app/models/solidus_afterpay/payment_method.rb:59]: I changed it back!

File: app/models/solidus_afterpay/payment_method.rb (hunk 2 above)
Comment (garciajordy) [reply app/models/solidus_afterpay/payment_method.rb:60]: We are using that in the `expires_in_seconds` method which will calculate the time until the cache gets expired.

Comment (ChristianRimondi) [reply app/models/solidus_afterpay/payment_method.rb:60]: yes, but you can just pass `1.day` to the `expires_in` argument. You don't need extra calculation

Comment (ChristianRimondi) [reply app/models/solidus_afterpay/payment_method.rb:60]: I mean, the gem will cache the value for 24 hours starting from the current time

File: app/models/solidus_afterpay/payment_method.rb (hunk 5 above)
Comment (ChristianRimondi) [reply app/models/solidus_afterpay/payment_method.rb:49]: NVM, this will call `configuration` even if the value is cached. We should find another way to handle it

File: app/models/solidus_afterpay/payment_method.rb (hunk 2 above)
Comment (garciajordy) [reply app/models/solidus_afterpay/payment_method.rb:60]: 👍

File: app/models/solidus_afterpay/payment_method.rb (hunk 5 above)
Comment (ChristianRimondi) [reply app/models/solidus_afterpay/payment_method.rb:49]:
```ruby
if Rails.cache.exist?(:solidus_afterpay_min_amount)
  Rails.cache.read(:solidus_afterpay_min_amount)
else
  Rails.cache.write(
    :solidus_afterpay_min_amount,
    configuration&.minimumAmount&.amount.to_f,
    expires_in: expires_in_seconds
  ) unless configuration.nil?

  configuration&.minimumAmount&.amount.to_f
end
```

Comment (ChristianRimondi) [reply app/models/solidus_afterpay/payment_method.rb:49]: I know it's ugly but it should work and we can work on it

File: spec/models/solidus_afterpay/payment_method_spec.rb (hunk 3 above)
Comment (garciajordy) [reply spec/models/solidus_afterpay/payment_method_spec.rb]: Done!

File: app/models/solidus_afterpay/payment_method.rb (hunk 4 above)
Comment (garciajordy) [reply app/models/solidus_afterpay/payment_method.rb:9]: Done!

File: spec/spec_helper.rb [hunk 6]
```diff
@@ -5,7 +5,7 @@
 
 # Run Coverage report
 require 'solidus_dev_support/rspec/coverage'
-
+require 'webmock/rspec'
```
Comment (vassalloandrea) [inline spec/spec_helper.rb:8]: I think that this is not needed anymore

File: lib/generators/solidus_afterpay/install/templates/initializer.rb [hunk 7]
```diff
@@ -2,4 +2,10 @@
 
 SolidusAfterpay.configure do |config|
   config.use_solidus_api = false
+
+  # There is a cache for retrieving the minimum amount, maximum amount and the currency from Afterpay
+  # the default value us set to 1.day, this is also recommended by Afterpay itself. If you prefer a shorter
+  # or longer time simply uncomment config.cache_expiry and add the preferred value.
+
```
Comment (vassalloandrea) [inline lib/generators/solidus_afterpay/install/templates/initializer.rb:9]: Remove this blank line, please since the comment above is related to the comment property below.

**Review by vassalloandrea — CHANGES_REQUESTED** [review]
Looks good! I left two minor comments

File: app/models/solidus_afterpay/payment_method.rb (hunk 5 above)
Comment (vassalloandrea) [reply app/models/solidus_afterpay/payment_method.rb]: I think that this global `fetch_configuration` method is missing @garciajordy

Comment (garciajordy) [reply app/models/solidus_afterpay/payment_method.rb]: Lol Didn't see that comment!

File: lib/generators/solidus_afterpay/install/templates/initializer.rb [hunk 8]
```diff
@@ -3,11 +3,8 @@
 SolidusAfterpay.configure do |config|
   config.use_solidus_api = false
 
-  # A class that extend SolidusAfterpay::BaseService, respond to .call, accepting a Spree::Order
-  # and return an Afterpay shipping rate object (check the Afterpay documentation)
-  # config.shipping_rate_builder_service_class = 'SolidusAfterpay::ShippingRateBuilderService'
-
-  # A class that extend SolidusAfterpay::BaseService, respond to .call, accepting a Spree::Order
-  # and return true or false
-  # config.update_order_attributes_service_class = 'SolidusAfterpay::UpdateOrderAttributesService'
```
Comment (ChristianRimondi) [inline lib/generators/solidus_afterpay/install/templates/initializer.rb:12]: Why did you remove these two comments? 🤔

File: spec/spec_helper.rb [hunk 9]
```diff
@@ -5,7 +5,6 @@
 
 # Run Coverage report
 require 'solidus_dev_support/rspec/coverage'
-
```
Comment (ChristianRimondi) [inline spec/spec_helper.rb:8]: Can you re-add the new line here?

**Review by ChristianRimondi — CHANGES_REQUESTED** [review]

File: lib/generators/solidus_afterpay/install/templates/initializer.rb (hunk 8 above)
Comment (garciajordy) [reply lib/generators/solidus_afterpay/install/templates/initializer.rb]: Done!

File: spec/spec_helper.rb (hunk 9 above)
Comment (garciajordy) [reply spec/spec_helper.rb]: Done!

**Review by ChristianRimondi — APPROVED** [review]
LGTM!

**Review by vassalloandrea — APPROVED** [review]
